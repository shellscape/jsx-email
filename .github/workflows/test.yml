name: Run Tests

on:
  workflow_dispatch:
  pull_request:
    types:
      - edited
      - opened
      - synchronize
  push:
    branches:
      - '*'
      - '!main'

jobs:
  validate:
    runs-on: ubuntu-latest
    name: Run Tests

    steps:
      - name: Checkout Commit
        uses: actions/checkout@v4
        with:
          fetch-depth: 10

      - name: Checkout Main
        run: |
          git fetch origin
          git branch -f main origin/main

      - name: Setup
        uses: ./.github/actions/setup

      - name: Build Projects
        run: |
          moon jsx-email:build
          moon create-jsx-email:build
          moon run :build --query "project~plugin-*"

      - name: Package Tests
        env:
          FORCE_COLOR: 1
        run: moon run :test --affected --concurrency 1 --remote
