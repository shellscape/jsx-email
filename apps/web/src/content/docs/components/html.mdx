---
title: HTML
description: Wraps the email document with a root element
type: component
---

import Install from '@components/mdx/install.mdx';

<h1>{frontmatter.title}</h1>
<p>{frontmatter.description}</p>

<Install type={frontmatter.type} />

## Usage

```jsx "Html"
// HtmlExample.jsx
import { Body, Head, Html } from 'jsx-email';

const Email = () => {
  return (
    <Html lang="en" dir="ltr">
      <Head>
        <title>Email Title</title>
      </Head>
      <Body>Email content</Body>
    </Html>
  );
};
```

:::note[VML Support]
By default, includes VML namespace attributes for Outlook compatibility.
:::

## Props

### Optional Props

#### dir

```ts
dir?: string;
```

Text direction for email content.

#### lang

```ts
lang?: string;
```

Language identifier for email content.

#### enableVML

```ts
enableVML?: boolean;
```

Controls VML namespace attributes. Defaults to `true`.

### Additional Props

This component expresses all of the [Common Component Props](https://react.dev/reference/react-dom/components/common) for `ComponentProps<'html'>`.
