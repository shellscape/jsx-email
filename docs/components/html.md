---
title: HTML
description: Wraps the email document with a root element
slug: html
type: component
---

<!--@include: @/include/header.md-->

<!--@include: @/include/install.md-->

## Usage

Add the component to your email template. Include styles where needed.

```jsx
import { Button, Html } from 'jsx-email';

const Email = () => {
  return (
    <Html lang="en" dir="ltr">
      <Button href="https://example.com" style={{ color: '#61dafb' }}>
        Click me
      </Button>
    </Html>
  );
};
```

## Component Props

```ts
lang?: string;
```

Identify the language of text content on the email

```ts
dir?: string;
```

Identify the direction of text content on the email

```ts
enableVML?: boolean;
```

This is used the `xmlns:o` and `xmlns:v` attributes to the `html` element for maximum VML compatibility
Default value is `true`

:::tip
This component also expresses all of the [Common Component Props](https://react.dev/reference/react-dom/components/common) for `ComponentProps<'html'>`.
:::
