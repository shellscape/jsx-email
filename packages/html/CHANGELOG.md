# @jsx-email/html ChangeLog

## v1.0.0

_2023-09-27_

### Breaking Changes

- chore!: initial release (032b213)