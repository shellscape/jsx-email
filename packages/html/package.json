{
  "name": "@jsx-email/html",
  "version": "1.0.1",
  "publishConfig": {
    "access": "public"
  },
  "description": "A JSX email component which wraps the email document with a root element",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/shellscape/jsx-email.git",
    "directory": "packages/html"
  },
  "homepage": "https://jsx.email/docs/components/html",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js"
      }
    },
    "./package.json": "./package.json"
  },
  "engines": {
    "node": ">=18.0.0"
  },
  "scripts": {
    "build": "tsup src/index.ts --format esm,cjs --dts --external react",
    "clean": "rm -rf dist",
    "format": "prettier --write \"**/*.{ts,tsx,md}\"",
    "format:check": "prettier --check \"**/*.{ts,tsx,md}\"",
    "test": "vi"
  },
  "files": [
    "dist/**"
  ],
  "keywords": [
    "react",
    "email"
  ],
  "peerDependencies": {
    "react": "^18.2.0"
  },
  "types": "dist/index.d.mts",
  "funding": {
    "type": "github",
    "url": "https://github.com/sponsors/shellscape"
  },
  "sideEffects": false
}
