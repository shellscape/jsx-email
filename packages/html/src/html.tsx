import React from 'react';

type RootProps = React.ComponentPropsWithoutRef<'html'>;

export interface HtmlProps extends RootProps {}

export const Html = ({
  children,
  lang = 'en',
  dir = 'ltr',
  ...props
}: React.PropsWithChildren<Readonly<HtmlProps>>) => (
  <html {...props} data-id="@jsx-email/html" lang={lang} dir={dir}>
    {children}
  </html>
);

Html.displayName = 'Html';
