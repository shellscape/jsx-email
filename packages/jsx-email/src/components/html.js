import { jsx as _jsx } from "react/jsx-runtime";
import { debug } from '../debug.js';
const debugProps = debug.elements.enabled ? { dataType: 'jsx-email/html' } : {};
export const Html = ({ children, lang = 'en', dir = 'ltr', enableVML = true, ...props }) => (_jsx("html", { ...props, ...debugProps, lang: lang, dir: dir, ...(enableVML
        ? {
            'xmlns:o': 'urn:schemas-microsoft-com:office:office',
            'xmlns:v': 'urn:schemas-microsoft-com:vml'
        }
        : {}), children: children }));
Html.displayName = 'Html';
//# sourceMappingURL=html.js.map