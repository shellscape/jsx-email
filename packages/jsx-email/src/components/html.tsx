import { debug } from '../debug.js';
import type { BaseProps, JsxEmailComponent } from '../types.js';

export interface HtmlProps extends BaseProps<'html'> {
  enableVML?: boolean;
}

const debugProps = debug.elements.enabled ? { dataType: 'jsx-email/html' } : {};

export const Html: JsxEmailComponent<HtmlProps> = ({
  children,
  lang = 'en',
  dir = 'ltr',
  enableVML = true,
  ...props
}) => (
  <html
    {...props}
    {...debugProps}
    lang={lang}
    dir={dir}
    {...(enableVML
      ? {
          'xmlns:o': 'urn:schemas-microsoft-com:office:office',
          'xmlns:v': 'urn:schemas-microsoft-com:vml'
        }
      : {})}
  >
    {children}
  </html>
);

Html.displayName = 'Html';
