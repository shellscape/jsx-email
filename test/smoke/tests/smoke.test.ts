/* eslint-disable no-await-in-loop */

import { test, expect } from '@playwright/test';

import { getHTML } from './helpers/html.js';

// TODO: Wire up tests for:
// - Mobile View
// - JSX View
// - HTML View
// - Copy and Download buttons on code views

const { error } = console;
const timeout = { timeout: 15e3 };

test('landing', async ({ page }) => {
  page.on('pageerror', (exception) => {
    console.log(`Uncaught exception: "${exception}"`);
    error(`Uncaught exception: "${exception}"`);
  });

  await page.goto('/');
  await expect(page).toHaveTitle('JSX email');

  page.getByText('JSX Email Preview');

  const landing = page.locator('#landing');

  expect(await landing.innerHTML({ timeout: 10e3 })).toMatchSnapshot();
  expect(page.locator('#Templates-link-Base')).toBeTruthy();
});

test('templates', async ({ page }) => {
  test.setTimeout(30e3);

  const allLinksSel = '#sidebar-tree a';
  const templatesButtonSel = '#Templates-sidebar-tree > button';
  const propsButtonSel = '#Props-sidebar-tree > button';

  await page.goto('/');
  await page.waitForSelector(templatesButtonSel, timeout);

  const templateButton = page.locator(templatesButtonSel);
  templateButton.click();

  await page.waitForSelector(templatesButtonSel, timeout);

  const propsButton = page.locator(propsButtonSel);
  propsButton.click();

  await page.waitForSelector(allLinksSel, timeout);

  const links = page.locator(allLinksSel);

  for (const link of await links.all()) {
    const name = await link.getAttribute('data-name');

    await test.step(`page: ${name}`, async () => {
      await link.click(timeout);

      const iframe = await page.frameLocator('#preview-frame');
      const html = await getHTML(iframe.locator('html'), { deep: true });

      expect(html).toMatchSnapshot({ name: `${name}.snap` });
    });
  }
});
