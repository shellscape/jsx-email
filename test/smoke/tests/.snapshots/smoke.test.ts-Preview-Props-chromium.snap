<html
  lang="en"
  dir="ltr"
>
  <head>
  </head>
  <body>
  </body>
</html>